src/construction/construction.py

```diff
@@ -1,282 +1,81 @@
-    write_csv(construction_output_filepath, constructed_df)
-    return constructed_df
-
-
-def apply_amendments(main_df, amendments_df):
-    """Apply amendments to the main snapshot."""
-    key_cols = ["reference", "year", "instance"]
-    numeric_cols = ["219", "220", "242", "243", "244", "245",
-                    "246", "247", "248", "249", "250"]
-    numeric_cols_new = [f"{i}_updated" for i in numeric_cols]
-
-    accepted_amendments_df = amendments_df.drop(amendments_df[~amendments_df.accept_changes].index)
-
-    if accepted_amendments_df.shape[0] == 0:
-        construction_logger.info("Amendments file contained no records marked for inclusion")
-        return main_df
-
-    # Drop the diff columns
-    accepted_amendments_df = accepted_amendments_df.drop(columns=[col for col in accepted_amendments_df.columns if col.endswith("_diff")])
-
-    # Join the amendments onto the main snapshot
-    amended_df = pd.merge(main_df, accepted_amendments_df, how="left", on=key_cols)
-
-    # Drop the old numeric cols and rename the amended cols
-    amended_df = amended_df.drop(columns=numeric_cols)
-    cols_to_rename = dict(zip(numeric_cols_new, numeric_cols))
-    amended_df = amended_df.rename(columns=cols_to_rename, errors="raise")
+    # Make a copy of the snapshot
+    updated_snapshot_df = snapshot_df.copy()
 
-    construction_logger.info(
-        f"{accepted_amendments_df.shape[0]} records amended during construction"
-    )
+    # Validate construction file and drop columns without constructed values
+    validate_data_with_schema(construction_df,
+                              "./config/construction_schema.toml")
+    construction_df = construction_df.dropna(axis="columns", how="all")
 
-    return amended_df
+    # Add flags to indicate whether a row was constructed
+    updated_snapshot_df["is_constructed"] = False
+    construction_df["is_constructed"] = True
 
+    # ! TEMPORARY Fix the year since the snapshot doesn't seem to have one?
+    updated_snapshot_df["year"] = 2022
 
-def apply_additions(main_df, additions_df):
-    """Apply additions to the main snapshot."""
-    # Drop records where accept_changes is False and if any remain, add them to main df
-    accepted_additions_df = additions_df.drop(
-        additions_df[~additions_df["accept_changes"]].index
-    )
-    if accepted_additions_df.shape[0] > 0:
-        added_df = pd.concat([main_df, accepted_additions_df], ignore_index=True)
-        construction_logger.info(
-            f"{accepted_additions_df.shape[0]} records added during construction"
-        )
-    else:
-        construction_logger.info("Additions file contained no records marked for inclusion")
-        return main_df
+    # Update the values with the constructed ones
+    construction_df.set_index(["reference", "year", "instance"], inplace=True)
+    updated_snapshot_df.set_index(["reference", "year", "instance"], inplace=True)
+    updated_snapshot_df.update(construction_df)
```

Comment:
This is so useful! Really glad to know how to do this - I knew about update() but was nervous using it... the "set_index" bit is obvs what's needed to make the thing work!